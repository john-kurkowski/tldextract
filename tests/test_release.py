"""Test the library maintainer release script."""

from __future__ import annotations

import dataclasses
import sys
from collections.abc import Iterator
from typing import Any
from unittest import mock

import pytest
from syrupy.assertion import SnapshotAssertion

from scripts import release


@dataclasses.dataclass
class Mocks:
    """Collection of all mocked objects used in the release script."""

    input: mock.Mock
    listdir: mock.Mock
    requests: mock.Mock
    subprocess: mock.Mock

    @property
    def mock_calls(self) -> dict[str, Any]:
        """A dict of _all_ calls to this class's mock objects."""
        return {
            k.name: getattr(self, k.name).mock_calls for k in dataclasses.fields(self)
        }


@pytest.fixture
def mocks() -> Iterator[Mocks]:
    """Stub network and subprocesses."""
    with mock.patch("builtins.input") as mock_input, mock.patch(
        "os.listdir"
    ) as mock_listdir, mock.patch("requests.post") as mock_requests, mock.patch(
        "subprocess.run"
    ) as mock_subprocess:
        yield Mocks(
            input=mock_input,
            listdir=mock_listdir,
            requests=mock_requests,
            subprocess=mock_subprocess,
        )


@pytest.mark.skipif(
    sys.platform == "win32", reason="Snapshot paths are different on Windows"
)
def test_happy_path(
    capsys: pytest.CaptureFixture[str],
    mocks: Mocks,
    monkeypatch: pytest.MonkeyPatch,
    snapshot: SnapshotAssertion,
) -> None:
    """Test the release script happy path.

    Simulate user input for a typical, existing release.

    This one test case covers most lines of the release script, without
    actually making network requests or running subprocesses. For an
    infrequently used script, this coverage is useful without being too brittle
    to change.
    """
    monkeypatch.setenv("GITHUB_TOKEN", "fake-token")

    mocks.input.side_effect = ["y", "5.0.1", "y"]

    mocks.listdir.return_value = ["archive1", "archive2", "archive3"]

    def mock_post(*args: Any, **kwargs: Any) -> mock.Mock:
        """Return _one_ response JSON that happens to match expectations for multiple requests."""
        return mock.Mock(
            json=mock.Mock(
                return_value={
                    "body": "Body start **Full Changelog**: fake-body",
                    "html_url": "https://github.com/path/to/release",
                }
            ),
        )

    mocks.requests.side_effect = mock_post

    mocks.subprocess.return_value.stdout = ""

    release.main()

    out, err = capsys.readouterr()

    assert mocks.mock_calls == snapshot
    assert out == snapshot
    assert err == snapshot
