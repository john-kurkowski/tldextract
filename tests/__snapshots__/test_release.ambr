# serializer version: 1
# name: test_happy_path
  dict({
    'input': _CallList([
      _Call(
        '',
        tuple(
          'Is this a test release? (y/n): ',
        ),
        dict({
        }),
      ),
      _Call(
        '',
        tuple(
          'Enter the version number: ',
        ),
        dict({
        }),
      ),
      _Call(
        '',
        tuple(
          'Does the build look correct? (y/n): ',
        ),
        dict({
        }),
      ),
    ]),
    'listdir': _CallList([
      _Call(
        '',
        tuple(
          'dist',
        ),
        dict({
        }),
      ),
    ]),
    'requests': _CallList([
      _Call(
        '',
        tuple(
          'https://api.github.com/repos/john-kurkowski/tldextract/releases/generate-notes',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/vnd.github+json',
            'Authorization': 'Bearer fake-token',
            'X-GitHub-Api-Version': '2022-11-28',
          }),
          'json': dict({
            'tag_name': '5.0.1',
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api.github.com/repos/john-kurkowski/tldextract/releases',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/vnd.github+json',
            'Authorization': 'Bearer fake-token',
            'X-GitHub-Api-Version': '2022-11-28',
          }),
          'json': dict({
            'body': '''
              * Bugfixes
                  * Indicate MD5 not used in a security context (FIPS compliance) ([#309](https://github.com/john-kurkowski/tldextract/issues/309))
              * Misc.
                  * Increase typecheck aggression
              
              **Full Changelog**: fake-body
            ''',
            'draft': True,
            'name': '5.0.1',
            'prerelease': False,
            'tag_name': '5.0.1',
          }),
        }),
      ),
    ]),
    'subprocess': _CallList([
      _Call(
        '',
        tuple(
          list([
            'git',
            'status',
            '--porcelain',
          ]),
        ),
        dict({
          'capture_output': True,
          'text': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'git',
            'tag',
            '-a',
            '5.0.1',
            '-m',
            '5.0.1',
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'rm',
            '-rf',
            PosixPath('dist'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'python',
            '-m',
            'build',
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'ls',
            '-l',
            PosixPath('dist'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'tar',
            'tvf',
            PosixPath('dist/archive1'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'tar',
            'tvf',
            PosixPath('dist/archive2'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'tar',
            'tvf',
            PosixPath('dist/archive3'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'twine',
            'upload',
            '--repository',
            'testpypi',
            PosixPath('dist/*'),
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
      _Call(
        '',
        tuple(
          list([
            'git',
            'push',
            '--tags',
            'origin',
            'master',
          ]),
        ),
        dict({
          'check': True,
        }),
      ),
    ]),
  })
# ---
# name: test_happy_path.1
  '''
  Version 5.0.1 tag added successfully.
  Previous dist folder removed successfully.
  Build created successfully.
  Contents of dist folder:
  Contents of tar files in dist folder:
  Build verified successfully.
  Release created successfully: https://github.com/path/to/release
  
  '''
# ---
# name: test_happy_path.2
  '''
  WARNING: dist folder contains incorrect number of files.
  
  '''
# ---
